lib/handler/file.c

```diff
@@ -40,51 +40,88 @@
 #include "h2o/http1.h"
 #include "../probes_.h"
 
+struct dsr_file_sender;
+typedef void (*dsr_file_sender_destroy_cb_t)(struct dsr_file_sender *sender);
+
 struct dsr_file_sender {
-    h2o_quic_ctx_t *ctx;
-    h2o_dsr_quic_packet_encryptor_t encryptor;
-    quicly_address_t dest_addr, src_addr;
+    int http_version;
     h2o_filecache_ref_t *file;
     h2o_socket_t *sock;
     uint64_t conn_id;
+    /**
+     * cleans up transport-specific fields in derived types
+     */
+    dsr_file_sender_destroy_cb_t on_destroy;
+};
+
+struct quic_dsr_file_sender {
+    struct dsr_file_sender super;
+    h2o_quic_ctx_t *ctx;
+    h2o_dsr_quic_packet_encryptor_t encryptor;
+    quicly_address_t dest_addr, src_addr;
 };
 
+static void quic_dsr_file_sender_on_destroy(struct dsr_file_sender *sender_)
+{
+    struct quic_dsr_file_sender *sender = H2O_STRUCT_FROM_MEMBER(struct quic_dsr_file_sender, super, sender_);
+    h2o_dsr_dispose_quic_packet_encryptor(&sender->encryptor);
+}
+
 static void destroy_dsr_file_sender(struct dsr_file_sender *sender)
 {
+    if (sender->on_destroy != NULL)
+        sender->on_destroy(sender);
     h2o_filecache_close_file(sender->file);
-    h2o_dsr_dispose_quic_packet_encryptor(&sender->encryptor);
     if (sender->sock != NULL)
         h2o_socket_close(sender->sock);
     free(sender);
 }
 
-static struct dsr_file_sender *create_dsr_file_sender(h2o_req_t *req, h2o_dsr_req_t *dsr_req, h2o_filecache_ref_t *file)
+static void init_dsr_file_sender_common(struct dsr_file_sender *sender, h2o_req_t *req, h2o_filecache_ref_t *file,
+                                        dsr_file_sender_destroy_cb_t on_destroy)
```

Comment:
I think we might want to retain `create_dsr_file_sender` that forces a particular way of instantiation (e.g., calling `h2o_mem_alloc`), considering the fact that `destroy_dsr_file_sender` assumes that the object has been instantiated in a particular way.

We can add a `size` argument to `create_dsr_file_sender`; see functions like `h2o_create_handler` for examples.

Reply:
I do not think I'm confused.

My point is that, assuming that we'd be requiring the file sender to be allocated on heap (by using `malloc`) and have a "destroy" function, we should be enforcing such design by providing a constructor that allocates memory and initializes the common fields.

That's the design pattern that we have for things like the handler; see https://github.com/h2o/h2o/blob/e0b48f11bb679660a82e1f64d7937337803a904f/lib/handler/file.c#L1024.

Reply:
Addressed in [7137cba](https://github.com/h2o/h2o/pull/3124/commits/7137cba451fc6fa7728b65382eb3445b1ea4afe2).